core/src/wallet/common/synchronizers/AbstractBlockchainExplorerAccountSynchronizer.h

```diff
@@ -322,52 +330,82 @@ namespace ledger {
+                                // WARNING: we have too many issues with that sync token because of blockchain explorer,
+                                // when we fail on reorg we try without sync token
+                                buddy->token = Option<void *>();
                             }
 
-                            //Update savedState's batches
-                            for (auto &batch : buddy->savedState.getValue().batches) {
-                                if (batch.blockHeight > lastBlockHeight) {
-                                    batch.blockHeight = (uint32_t)lastBlockHeight;
-                                    batch.blockHash = lastBlockHash;
+                            //Get its block/block height
+                            auto &failedBatch = buddy->savedState.getValue().batches[currentBatchIndex];
+                            auto const failedBlockHeight = failedBatch.blockHeight;
+                            auto const failedBlockHash = failedBatch.blockHash;
+
+                            if (failedBlockHeight > 0) {
+                                //Delete data related to failedBlock (and all blocks above it)
+                                buddy->logger->info("Deleting blocks above block height: {}", failedBlockHeight);
+
+                                soci::session sql(buddy->wallet->getDatabase()->getPool());
+                                sql << "DELETE FROM blocks where height >= :failedBlockHeight", soci::use(
+                                        failedBlockHeight);
+
+                                //Get last block not part from reorg
+                                auto lastBlock = BlockDatabaseHelper::getLastBlock(sql,
+                                                                                   buddy->wallet->getCurrency().name);
+
+                                //Resync from the "beginning" if no last block in DB
+                                int64_t lastBlockHeight = 0;
+                                std::string lastBlockHash;
+                                if (lastBlock.nonEmpty()) {
+                                    lastBlockHeight = lastBlock.getValue().height;
+                                    lastBlockHash = lastBlock.getValue().blockHash;
                                 }
-                            }
 
-                            //Save new savedState
-                            buddy->preferences->editor()->template putObject<BlockchainExplorerAccountSynchronizationSavedState>(
-                                    "state", buddy->savedState.getValue())->commit();
+                                //Update savedState's batches
+                                for (auto &batch : buddy->savedState.getValue().batches) {
+                                    if (batch.blockHeight > lastBlockHeight) {
+                                        batch.blockHeight = (uint32_t) lastBlockHeight;
```

Comment:
We are performing a narrow cast which can result to an unexpected value.

I would like to introduce you `gsl` header-only library. `gsl::narrow_cast` performs some check during the cast and throws a `narrowing_error` if the resulting value is different from the uncasted value.

```suggestion
                                        batch.blockHeight = static_cast<uint32_t>(lastBlockHeight);
```

Reply:
Really cool for `gsl` !! 
As discussed f2f, this is not something changed by this PR, Github is clearly messed up. but still the point is valid and should be treated in a dedicated PR for all our code ! I will create the corresponding ticket.